libflux/src/semantic/nodes.rs

```diff
@@ -149,12 +217,35 @@ pub struct ReturnStmt {
 
 #[derive(Debug, PartialEq, Clone)]
 pub struct VariableAssgn {
+    free: Tvars,
+    cons: Option<TvarKinds>,
+
     pub loc: ast::SourceLocation,
 
     pub id: Identifier,
     pub init: Expression,
 }
 
+impl VariableAssgn {
+    pub fn new(id: Identifier, init: Expression, loc: ast::SourceLocation) -> VariableAssgn {
+        VariableAssgn {
+            free: Tvars::new(),
+            cons: None,
+            loc,
+            id,
+            init,
+        }
+    }
+}
+
+pub fn poly_type_of(v: &VariableAssgn) -> PolyType {
+    PolyType {
+        free: v.free.clone(),
+        cons: v.cons.clone(),
```

Comment:
I think we tend to `clone` too much. Why not returning pointers and let the callers decide to clone if necessary?

It is weird to see a getter method that returns an owned object instead of a pointer 🤔

Reply:
I think that's a good rule to follow in general. If the method takes a reference then it should return a reference. The caller can decide if it needs to clone or not. I will update these methods.

Reply:
Actually @affo in this case I think I need to return an owned object. The reason is because I only attach the free variables to a variable assignment node. `poly_type_of` then constructs a polytype using those free variables and the monotype of the `init` field.